### PR TITLE
[Fix] `shallow`: `.setState()`: stub out `setState` on non-root code paths as well.

### LINKED ISSUE #1756: `this` is undefined in bound functions when shallow rendering
**Describe the bug**
The following test passes in 3.4.1 but fails in 3.4.2. 
```js
class Comp extends Component {
  state = {
    key: "",
  }

  componentDidMount() {
    this.instanceFunction();
  }

  instanceFunction = () => this.setState(() => ({ key: "value" }));

  render() {
    const { key } = this.state;
    return null;
  }
}

it("should work", () => {
  shallow(<Comp />);
});
```

The error in 3.4.2 is:
```shell
FAIL src/test.js
  ✕ should work (16ms)

  ● should work

    Method “setState” is only meant to be run on a single node. undefined found instead.

      11 |   }
      12 |
    > 13 |   instanceFunction = () => this.setState(() => ({ key: "value" }));
         |                                 ^
      14 |
      15 |   render() {
      16 |     const { key } = this.state;

      at ShallowWrapper.single (node_modules/enzyme/build/ShallowWrapper.js:1718:17)
      at ShallowWrapper.setState (node_modules/enzyme/build/ShallowWrapper.js:499:14)
      at Comp.ShallowWrapper.instance.setState (node_modules/enzyme/build/ShallowWrapper.js:184:33)
      at Comp.setState [as instanceFunction] (src/test.js:13:33)
      at Comp.instanceFunction [as componentDidMount] (src/test.js:10:10)
      at node_modules/enzyme/build/ShallowWrapper.js:189:20
      at Object.batchedUpdates (node_modules/enzyme-adapter-react-16/build/ReactSixteenAdapter.js:392:22)
      at new ShallowWrapper (node_modules/enzyme/build/ShallowWrapper.js:188:24)
      at shallow (node_modules/enzyme/build/shallow.js:21:10)
      at Object.<anonymous> (src/test.js:22:3)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        2.147s
Ran all test suites matching /src\/test.js/i.
error Command failed with exit code 1.
```

**To Reproduce**
1. Run the above test with `enzyme@3.4.2`

**Expected behavior**
The test should pass.

**Desktop (please complete the following information):**
 - OS: Mac OSX (but also fails in Windows)
 - Browser n/a
 - Version n/a

### PR DESCRIPTION
**Pull request by ljharb**
Fixes #1756

cc @koba04

### CODE REVIEW

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 1]
```diff
@@ -2233,6 +2233,31 @@ describe('shallow', () => {
...
+          this.instanceFunction();
+        }
+
+        render() {
+          const { key } = this.state;
+          return key ? null : null;
+        }
+      }
+
+      expect(shallow(<Comp />).debug()).to.equal('');
```
Comment (koba04) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: Does `setState` in `componentDidMount` invoke `componentDidUpdate`?

File: packages/enzyme/src/ShallowWrapper.js [hunk 2]
```diff
@@ -195,6 +193,15 @@ class ShallowWrapper {
       privateSetNodes(this, nodes);
     }
     privateSet(this, OPTIONS, root ? root[OPTIONS] : options);
+
+    if (!instance) {
+      ({ instance } = this[RENDERER].getNode());
+    }
+    // Ensure to call componentDidUpdate when instance.setState is called
+    if (instance && lifecycles.componentDidUpdate.onSetState && !instance[SET_STATE]) {
```
Comment (koba04) [inline packages/enzyme/src/ShallowWrapper.js]: Nits: I should have overridden `setState` only in `options.disableLifecycleMethods` is false.

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx (hunk 1 above)
Comment (ljharb) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: it should; cDU is invoked whenever it updates.

File: packages/enzyme/src/ShallowWrapper.js (hunk 2 above)
Comment (ljharb) [reply packages/enzyme/src/ShallowWrapper.js]: i'll add a test for that.

Comment (ljharb) [reply packages/enzyme/src/ShallowWrapper.js]: actually, can you write one? :-D i'm not sure where to put it.

or is this just an optimization?

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx (hunk 1 above)
Comment (koba04) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: In order to be called cDU, we should stub `setState` so I think we should move calling cDM after the stubbing.

File: packages/enzyme/src/ShallowWrapper.js (hunk 2 above)
Comment (koba04) [reply packages/enzyme/src/ShallowWrapper.js]: https://github.com/airbnb/enzyme/pull/1763/commits/a05d4cb5dcd84dfd0ba3c8f074641b173e85242f is the right place. It is unnecessary when `options. disableLifecycleMethods` is `true`.

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx (hunk 1 above)
Comment (koba04) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: I'll write the tests and the patch.
https://github.com/airbnb/enzyme/compare/fix_shallow_setstate...koba04:fix_shallow_setstate?expand=1

**Comment by ljharb** [comment]
Thanks, this looks great. Will merge once tests pass.

**Comment by monken** [comment]
Not sure why, but the 3.4.3 release broke my tests. I receive `Error: ShallowWrapper::setState() can only be called on the root` errors all over the place.

The relevant test is attached. I was able to figure out that `componentDidMount` is called twice with the 3.4.3 release, which was not the case before.

```js
  it('<MyComponent />', async () => {
    fetch.mockResponse(JSON.stringify(jsonData));
    const wrapper = shallow(<MyComponent />, { disableLifecycleMethods: true });
    await wrapper.instance().componentDidMount();
    wrapper.update();
    expect(fetch.mock.calls.length).toEqual(1);
    expect(wrapper.find(Table).length).toEqual(1);
  });
```

**Comment by ljharb** [comment]
@monken please file a new issue; commenting on a merged PR isn't going to get anything fixed.

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 3]
```diff
@@ -2142,7 +2142,7 @@ describe('shallow', () => {
       expect(wrapper.find('.bar')).to.have.lengthOf(1);
     });
 
-    it.skip('allows setState inside of componentDidMount', () => {
+    it('allows setState inside of componentDidMount', () => {
```
Comment (talco) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: should not allow setState inside of componentDidMount

Comment (ljharb) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: Why would it not?

**Pull request merged**